Automattic/newspack-blocks
# feat: redesign modal checkout

### All Submissions:






### Changes proposed in this Pull Request:

1205234045751551-as-1205933702651188

Implements a redesigned and new 2-step strategy for the modal checkout.

The modal checkout class no longer performs form validation, it now uses `woocommerce_checkout_update_totals` so  [it validates through Woo without having to place the order](https://github.com/woocommerce/woocommerce/blob/798e79f7b1bb5cd3db376de8712ad122d0094112/plugins/woocommerce/includes/class-wc-checkout.php#L1262). By validating through Woo, the modal checkout should natively support any custom field combination Woo and its extensions may use. This new strategy also uses XHR instead of a regular form submission, which should provide a better and faster experience for the reader.

To also increase support for core Woo checkout functionality, the functional aspects of the modal checkout were moved from the template file to [hooks](https://github.com/Automattic/newspack-blocks/blob/7fd2af31749c8570148b7a0c5294a2443e0e8014/includes/class-modal-checkout.php#L52-L54). This allows the template file to be modified/updated without impacting the requirements that allow the checkout to function in a modal. The new `src/modal-checkout/templates/form-checkout.php` is now intentionally very similar to Woo's original template file.

A significant part of the work here has been around `src/modal-checkout/index.js`, to allow the new 2-step approach to work as the design proposes. Some important parts are:

 - When in the first step (editing billing/shipping fields), [the original form event handlers are replaced](https://github.com/Automattic/newspack-blocks/blob/635a685a22878fde6649cc4cb2aef8b96e93ef5d/src/modal-checkout/index.js#L180-L185) so the form submission performs [our ajax request](https://github.com/Automattic/newspack-blocks/blob/635a685a22878fde6649cc4cb2aef8b96e93ef5d/src/modal-checkout/index.js#L286) (which adds the `woocommerce_checkout_update_totals` parameter).
 - [A custom error handler](https://github.com/Automattic/newspack-blocks/blob/7fd2af31749c8570148b7a0c5294a2443e0e8014/src/modal-checkout/index.js#L84-L89) for the 1st step identifies whether the incoming errors are field-related or generic and appends the error message to the field or at the top accordingly.

### Known issues

- The coupon form is inconsistent. When applying a coupon the form disappears, which can be considered fine because it reappears after removing the coupon. It's not ok when applying an inexistent coupon, which shows you an error message and the coupon form disappears, becoming irrecoverable. (PR: https://github.com/Automattic/newspack-blocks/pull/1629)
- Linter is accusing a CSS kebab-case issue even though the line has `/* stylelint-disable-line */`. Not sure how to deal with this.

### Subscription

<img width="647" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/87c99af8-4153-455b-a9e5-98e4a8c5235b">

<img width="638" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/9e1d0d58-2488-4777-8032-6b55122097a7">

### One-time donation

<img width="661" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/e8017d85-1156-4378-a85b-47161b714777">

### Recurring donation

<img width="651" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/b2da7fe1-0e14-419b-bff1-a4a877ac7604">

### Recurring donation (known customer with known card)

<img width="651" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/629a198a-ea16-4bb4-83c0-cc5882f1ea60">

### Shippable product

<img width="649" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/f7a2e5b6-ed99-4492-806c-0b9cb193c9b8">

### Field validation

<img width="641" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/202d7b70-f5f7-4a96-9b33-a1e8b84ab25a">

### How to test the changes in this Pull Request:

1. Checkout this branch and https://github.com/Automattic/newspack-plugin/pull/2805
2. If you don't have any, setup a shipping zone in Woo:
    1. Visit WooCommerce -> Settings -> Shipping
    2. Click "Add shipping zone"
    3. Give it a name (the country name)
    4. Set the "Zone region" to your country
    5. Click "Add shipping method" and leave it as "Flat rate"
    6. Save changes
3. Make sure you have coupons enabled and have a valid coupon
7. Create a product that can be shipped
8. Create a regular custom subscription
9. Create a page with a Checkout Button block for each of the created products
10. Also add a Donate Block to the page
11. Visit the page, click to buy the shippable product
12. Confirm billing and shipping fields render as expected
13. Attempt to "Continue" without entering all required fields
14. Confirm validation errors render for each field
15. Enter a zipcode that you know to be an invalid format
16. Confirm it errors with "Billing ZIP Code is not a valid postcode / ZIP."
17. Now enter valid fields and confirm you are able to go to the next step
18. Confirm the second step renders a summary of the "Billing details" and "Shipping details"
19. Click "Back" and confirm you are able to edit the fields again
20. Scroll down and toggle "Ship to a different address?"
21. Modify the address and click "Continue"
22. Confirm the summary reflects the different address
23. Apply the coupon and confirm the order review table renders with the coupon information
24. Click to remove the coupon and confirm the order review table is no longer visible
25. Re-apply the coupon so we can confirm the coupon persists while placing the order
26. Submit and confirm the order is placed with the applied coupon
27. Leave the modal and click to purchase the custom subscription
28. Confirm the billing fields reflect the selection from "Newspack -> Reader Revenue"
29. Go through the entire flow and confirm the order is placed without issues
30. Leave the modal and click to donate
31. In the second step, confirm the coupon form is not available
32. Confirm the button to place the order says "Donate now" instead of "Complete transaction"
33. Click to donate and confirm the order is placed

### Other information:
